CefSharp.Core/Internals/ClientAdapter.cpp

```diff
@@ -323,15 +350,36 @@ namespace CefSharp
 
         bool ClientAdapter::OnBeforeBrowse(CefRefPtr<CefBrowser> browser, CefRefPtr<CefFrame> frame, CefRefPtr<CefRequest> request, bool isRedirect)
         {
-            IRequestHandler^ handler = _browserControl->RequestHandler;
-            if (handler == nullptr)
+            if (browser->IsPopup())
             {
-                return false;
+                auto popupHandler = _browserControl->PopupHandler;
+                if (popupHandler != nullptr)
+                {
+                    auto browserId = browser->GetIdentifier();
+                    IBrowser^ entry;
+                    if (_popupBrowsers->TryGetValue(browserId, entry))
+                    {
+                        auto frameWrapper = gcnew CefFrameWrapper(frame, _browserAdapter);
+                        CefRequestWrapper^ wrapper = gcnew CefRequestWrapper(request);
+                        return popupHandler->OnBeforeBrowse(_browserControl, entry, wrapper, isRedirect, gcnew CefFrameWrapper(frame, _browserAdapter));
```

Comment:
> For  CefBrowser  it's pretty clear and I think that's pretty straight forward to reuse the wrapper. For  CefFrame , I'm not sure that there is such a notification.

Well, there's good news and bad news. There seems to be a notification, but it's only in the rendering process as best I can see looking through the .h files: `CefRenderProcessHandler::OnContextReleased`.

Ugh.

> What does the extra layer get us in this instance? If we Have a reference to the wrapper and simply release the  unamanged  resource, is there anything more required? (I'm not opposed to the idea, I'm just unsure what the extra overhead gets us)

I suppose as long as we have some kind of: `IsDisposed` property on the object, the effect would be the same. We'd set it to `true` when CEF tells us the underlying thing is no longer useful. Then we would add all of those fun `ThrowIfDisposed` calls to the wrappers. Whee! :smilie:

Bill